File: ts/src/verification.ts
// The independent verifier: re-checks a proof-carrying answer using only the
// embedded public key and proofs.

import { createHash } from "node:crypto";

import { verifyInclusion, type ProofStep } from "./merkle.ts";
import type {
  ChunkResult,
  ProofStepModel,
  QueryResponse,
  VerificationResult,
} from "./models.ts";
import { verifyRoot } from "./signing.ts";

/**
 * Verify a QueryResponse, proving each cited chunk is committed under the
 * signed root without trusting the server that produced the answer. Throws if a
 * proof references a chunk id with no matching citation.
 */
export function verifyResponse(resp: QueryResponse): VerificationResult {
  const sr = resp.signed_root;
  const publicKey = Buffer.from(sr.public_key, "hex");
  const root = Buffer.from(sr.root, "hex");
  const prev = sr.prev_root ? Buffer.from(sr.prev_root, "hex") : null;
  const signature = Buffer.from(sr.signature, "hex");

  const signatureOk = verifyRoot(publicKey, signature, root, prev, sr.leaf_count);
  let allOk = signatureOk;

  const chunks: ChunkResult[] = [];
  for (const proof of resp.proofs) {
    const leafBytes = leafFor(resp, proof.chunk_id);
    // Integrity: the recorded hash is the PLAIN SHA-256 of the chunk text, not
    // the 0x00-prefixed leaf hash used inside the Merkle tree.
    const recomputed = createHash("sha256").update(leafBytes).digest("hex");
    const hashOk = recomputed === proof.sha256;
    const path = decodePath(proof.merkle_path);
    const inclusionOk = verifyInclusion(leafBytes, path, root);
    allOk = allOk && hashOk && inclusionOk;
    chunks.push({ chunk_id: proof.chunk_id, hash_ok: hashOk, inclusion_ok: inclusionOk });
  }
  return { verified: allOk, signature_ok: signatureOk, chunks };
}

/** Convert wire proof steps (hex siblings) into raw-byte proof steps. */
function decodePath(steps: ProofStepModel[]): ProofStep[] {
  return steps.map((step) => ({
    sibling: Buffer.from(step.sibling, "hex"),
    side: step.side === "left" ? "left" : "right",
  }));
}

/** Return the UTF-8 bytes of the citation text for a chunk id. */
function leafFor(resp: QueryResponse, chunkId: string): Buffer {
  const citation = resp.citations.find((c) => c.chunk_id === chunkId);
  if (!citation) {
    throw new Error(`no citation text for chunk ${chunkId}`);
  }
  return Buffer.from(citation.text, "utf8");
}

File: rust/src/verification.rs
//! The independent verifier: re-checks a proof-carrying answer using only the
//! embedded public key and proofs.

use sha2::{Digest, Sha256};

use crate::merkle::{verify_inclusion, ProofStep};
use crate::models::{ChunkResult, ProofStepModel, QueryResponse, VerificationResult};
use crate::signing::verify_root;

/// Verify a [`QueryResponse`], proving each cited chunk is committed under the
/// signed root without trusting the server that produced the answer.
pub fn verify_response(resp: &QueryResponse) -> Result<VerificationResult, String> {
    let sr = &resp.signed_root;
    let public_key = hex::decode(&sr.public_key).map_err(|e| format!("bad public_key hex: {e}"))?;
    let root = hex::decode(&sr.root).map_err(|e| format!("bad root hex: {e}"))?;
    let prev: Option<Vec<u8>> = match &sr.prev_root {
        Some(s) if !s.is_empty() => {
            Some(hex::decode(s).map_err(|e| format!("bad prev_root hex: {e}"))?)
        }
        _ => None,
    };
    let signature = hex::decode(&sr.signature).map_err(|e| format!("bad signature hex: {e}"))?;

    let sig_ok = verify_root(
        &public_key,
        &signature,
        &root,
        prev.as_deref(),
        sr.leaf_count,
    );
    let mut all_ok = sig_ok;

    let mut chunks = Vec::with_capacity(resp.proofs.len());
    for proof in &resp.proofs {
        let leaf_bytes = leaf_for(resp, &proof.chunk_id)?;
        // Integrity: the recorded hash is the PLAIN SHA-256 of the chunk text,
        // not the 0x00-prefixed leaf hash used inside the Merkle tree.
        let recomputed = hex::encode(Sha256::digest(&leaf_bytes));
        let hash_ok = recomputed == proof.sha256;
        let path = decode_path(&proof.merkle_path)?;
        let inclusion_ok = verify_inclusion(&leaf_bytes, &path, &root);
        all_ok = all_ok && hash_ok && inclusion_ok;
        chunks.push(ChunkResult {
            chunk_id: proof.chunk_id.clone(),
            hash_ok,
            inclusion_ok,
        });
    }
    Ok(VerificationResult {
        verified: all_ok,
        signature_ok: sig_ok,
        chunks,
    })
}

/// Convert wire proof steps (hex siblings) into raw-byte [`ProofStep`]s.
fn decode_path(steps: &[ProofStepModel]) -> Result<Vec<ProofStep>, String> {
    steps
        .iter()
        .map(|s| {
            let sibling = hex::decode(&s.sibling).map_err(|e| format!("bad sibling hex: {e}"))?;
            Ok(ProofStep {
                sibling,
                side: s.side.clone(),
            })
        })
        .collect()
}

/// Return the UTF-8 bytes of the citation text for a chunk id.
fn leaf_for(resp: &QueryResponse, chunk_id: &str) -> Result<Vec<u8>, String> {
    resp.citations
        .iter()
        .find(|c| c.chunk_id == chunk_id)
        .map(|c| c.text.as_bytes().to_vec())
        .ok_or_else(|| format!("no citation text for chunk {chunk_id}"))
}
